Rationale.md

```diff
@@ -0,0 +1,198 @@
+some inefficiencies.
+
+Local values can also be pre-colored, meaning that multiple incoming SSA values
+which have separate liveness can "share" the storage represented by a
+local. This offloads some of the optimization work from the WebAssembly VM.
+
+
+## Variable-Length Argument Lists
+
+C and C++ compilers are expected to implement variable-length argument lists by
+storing arguments in a buffer in linear memory and passing a pointer to the
+buffer. This greatly simplifies WebAssembly VM implementations by punting this
+ABI consideration to the front-end compiler. It does negatively impact
+performance, but variable-length calls are already somewhat slow.
+
+
+## Multiple Return Values
+
+TODO
+
+
+## Indirect Calls
+
+The exact semantics of indirect function calls, function pointers, and what
+happens when calling the wrong function, are still being discussed.
+
+Fundamentally linear memory is a simple collection of bytes, which means that
+some integral representation of function pointers must exist. It's desirable to
+hide the actual address of generated code from untrusted code because that would
+be an unfortunate information leak which could have negative security
+implications. Indirection is therefore desired.
+
+One extra concern is that existing C++ code sometimes stores data inside of what
+is usually a function pointer. This is expected to keep working.
+
+Dynamic linking further complicates this: WebAssembly cannot simply standardize
+on fixed-size function tables since dynamically linked code can add new
+functions, as well as remove them.
+
+
+## Expressions with Control Flow
+
+Expression trees offer significant size reduction by avoiding the need for
+`set_local`/`get_local` pairs in the common case of an expression with only one,
+immediate use. The following primitives provide AST nodes that express control
+flow and thus allow more opportunities to build bigger expression trees and
+further reduce `set_local`/`get_local` usage (which constitute 30-40% of total
+bytes in the
+[polyfill prototype](https://github.com/WebAssembly/polyfill-prototype-1)).
+Additionally, these primitives are useful building blocks for
+WebAssembly-generators (including the JavaScript polyfill prototype).
+
+
+## Limited Local Nondeterminism
+
+The current specification is fairly strict when it comes to
+[limited local nondeterminism](Nondeterminism.md) of operations: it tries to
+specify all possible corner cases.
+
+As WebAssembly gets implemented and tested with multiple languages on multiple
+architectures there may be a need to revisit some of the decisions:
+
+* When all relevant hardware implement features the same way then there's no
+  need to get creative.
+* When different languages have different expectations then it's unfortunate if
+  WebAssembly measurably penalizes one's performance by enforcing determinism
+  which that language doesn't care about, but which another language may want.
```

Comment:
I'm not sure that's quite true: if it were the case we'd avoid specifying e.g. shift and divide the way we do! I'm really hoping that we start of with something that seems good, and before MVP we adjust what we have in preparation for other languages.